File: bridge/DynamicsCompressorNode.cpp
#include "dart_api/dart_api.h"
#include "LabSound/LabSound.h"
#include "KeepNode.cpp"
using namespace lab;

DART_EXPORT int createDynamicsCompressorNode(AudioContext* context) {
    auto node = std::make_shared<DynamicsCompressorNode>(*context);
    return keepNode(node);
}

DART_EXPORT int DynamicsCompressorNode_threshold(int nodeId) {
    auto node = std::static_pointer_cast<DynamicsCompressorNode>(getNode(nodeId));
    return node ? keepAudioParam(nodeId, 1, node->threshold()) : -1;
}

DART_EXPORT int DynamicsCompressorNode_knee(int nodeId) {
    auto node = std::static_pointer_cast<DynamicsCompressorNode>(getNode(nodeId));
    return node ? keepAudioParam(nodeId, 2, node->knee()) : -1;
}

DART_EXPORT int DynamicsCompressorNode_ratio(int nodeId) {
    auto node = std::static_pointer_cast<DynamicsCompressorNode>(getNode(nodeId));
    return node ? keepAudioParam(nodeId, 3, node->ratio()) : -1;
}

DART_EXPORT int DynamicsCompressorNode_attack(int nodeId) {
    auto node = std::static_pointer_cast<DynamicsCompressorNode>(getNode(nodeId));
    return node ? keepAudioParam(nodeId, 4, node->attack()) : -1;
}

DART_EXPORT int DynamicsCompressorNode_release(int nodeId) {
    auto node = std::static_pointer_cast<DynamicsCompressorNode>(getNode(nodeId));
    return node ? keepAudioParam(nodeId, 5, node->release()) : -1;
}

DART_EXPORT int DynamicsCompressorNode_reduction(int nodeId) {
    auto node = std::static_pointer_cast<DynamicsCompressorNode>(getNode(nodeId));
    return node ? keepAudioParam(nodeId, 6, node->reduction()) : -1;
}

File: lab_sound_ffi/lib/src/lab_sound/dynamics_compressor_node.dart
import 'audio_context.dart';
import 'audio_node.dart';
import 'audio_param.dart';
import 'lab_sound.dart';

class DynamicsCompressorNode extends AudioNode {

  late AudioParam threshold;
  late AudioParam knee;
  late AudioParam ratio;
  late AudioParam attack;
  late AudioParam release;

  late AudioParam reduction;

  DynamicsCompressorNode(AudioContext ctx): super(ctx, LabSound().createDynamicsCompressorNode(ctx.pointer)) {
    threshold = AudioParam(this.ctx, nodeId, LabSound().DynamicsCompressorNode_threshold(nodeId));
    knee = AudioParam(this.ctx, nodeId, LabSound().DynamicsCompressorNode_knee(nodeId));
    ratio = AudioParam(this.ctx, nodeId, LabSound().DynamicsCompressorNode_ratio(nodeId));
    attack = AudioParam(this.ctx, nodeId, LabSound().DynamicsCompressorNode_attack(nodeId));
    release = AudioParam(this.ctx, nodeId, LabSound().DynamicsCompressorNode_release(nodeId));
    reduction = AudioParam(this.ctx, nodeId, LabSound().DynamicsCompressorNode_reduction(nodeId));
  }

}